brighthive/data-resource-api
# Feat/put patch many to many

## Overview

Adds PUT and PATCH to the many to many routes and allows you to enable and secure the routes.

Refactors the many to many tests.

### Checklist


### Notes

This adds two utility functions, `error_if_resource_is_disabled` and `is_secured`. There is a problem with these implementations. If you have more than one many to many resource and the first resource uses the wrong values for 'methods' then all many to many looks ups will probably fail and raise an error. The solutions to this is validating that the descriptor is correct or refactoring these functions to not fail in that way.

## Testing Instructions
1. Test many to many PUT and PATCH

 * Start after checking out this branch
 * Build the image with `docker build -t brighthive/data-resource-api:1.0.0-alpha .`
 * Run `docker-compose up`
 * Give the application a minute or two to spin up
 * `POST` three credentials to `http://localhost:8000/credentials`
```
{
	"credential_name": "credential test"
}
```

 * `POST` a program with `'credentials': [1,2]` to `http://localhost:8000/programs`
```
{
	"program_name": "aprogram 1",
	"program_code": 1,
	"program_description": ".",
	"program_status": ".",
	"program_fees": 1,
	"eligibility_criteria": ".",
	"program_url": ".",
	"credentials": [1,2]
}
```

 * `PUT` the following to `http://localhost:8000/programs/1/credentials`. It should return `'credentials': [2,3]`
```
{
	"credentials": [2,3]
}
```

 * `PATCH` thw following to `http://localhost:8000/programs/1/credentials'. It should return `'credentials': [2,3,1]`
```
{
	"credentials": [1,1,2]
}
```

2. Test enabled
 * Open the schema descriptor file `programs`. Under `custom` set any of the methods to `'enabled': false`.
 * Perform the operation and you should receive a method not allowed response.

3. Test secured
 * Open the schema descriptor file 'programs'. Under `custom` set any of the methods to `'secured': true` and `'enabled': true`.
 * Perform the operation without auth tokens and you should receive an unauthorized error.
 * Acquire auth tokens and perform the operation. You should successfully perform the operation.


Handles https://github.com/brighthive/data-resource-api/issues/6
